### PR TITLE
Expose EncodedSemanticClassificationsRequest in protocol.d.ts

### LINKED ISSUE #42587: Add semantic highlighting API to Protocol
Follow up on #41262

It looks like the `encodedSemanticClassifications-full` types didn't make it into the `protocol.d.ts` (I checked 4.2.0-dev.20210201)

The apis seem to work but the types are missing 

/cc @orta

### PR DESCRIPTION
**Pull request by orta**
Fixes #42587 - I think because this request used to have the same args as many other requests then it was not included in the `protocol.d.ts`. 

Now it differs, so I've removed the internal markers 👍🏻

### CODE REVIEW

**Review by sheetalkamat — APPROVED** [review]

File: tests/baselines/reference/api/tsserverlibrary.d.ts [hunk 1]
```diff
@@ -7133,6 +7133,30 @@ declare namespace ts.server.protocol {
          */
         body?: string[];
     }
+    /**
+     * A request to get encoded semantic classifications for a span in the file
+     */
+    interface EncodedSemanticClassificationsRequest extends FileRequest {
```
Comment (mjbvz) [inline tests/baselines/reference/api/tsserverlibrary.d.ts:7139]: Do we also need to add the response types? We currently have to maintain a copy of them in VS Code: https://github.com/microsoft/vscode/blob/3d500ebd8b4a919e8631e916f417e5c68eb90d3f/extensions/typescript-language-features/src/languageFeatures/semanticTokens.ts#L275

Comment (sheetalkamat) [reply tests/baselines/reference/api/tsserverlibrary.d.ts:7139]: Aren’t they in typescript/tsserver .d.ts ?

Comment (mjbvz) [reply tests/baselines/reference/api/tsserverlibrary.d.ts:7139]: We currently only use `protocol.d.ts` with VS Code. I think this should have all the types used communicating from or to tsserver

Comment (orta) [reply tests/baselines/reference/api/tsserverlibrary.d.ts:7139]: Makes sense, added

File: src/server/protocol.ts [hunk 2]
```diff
@@ -868,6 +866,14 @@ namespace ts.server.protocol {
         format?: "original" | "2020"
     }
 
+    /** The response for a EncodedSemanticClassificationsRequest */
+    export interface EncodedSemanticClassificationsResponse extends Response {
+        body?: {
```
Comment (sheetalkamat) [inline src/server/protocol.ts:871]: Pull this out as `EncodedSemanticClassificationsResponseBody` interface

File: src/server/protocol.ts [hunk 3]
```diff
@@ -868,6 +866,14 @@ namespace ts.server.protocol {
         format?: "original" | "2020"
     }
 
+    /** The response for a EncodedSemanticClassificationsRequest */
+    export interface EncodedSemanticClassificationsResponse extends Response {
+        body?: {
+            endOfLineState: EndOfLineState;
+            spans: number[];
```
Comment (sheetalkamat) [inline src/server/protocol.ts:873]: Dont we need specific enum for  classifications returned otherwise vscode would need separate list?

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

**Comment by mjbvz** [comment]
Hi, just want to check if this is on track for TS 4.2.1. If not, could we please get this  into a recovery build of TS 4.2 instead of waiting for TS 4.3? This would let us better validate the semantic highlighting support API while switching to the native TS implementation (https://github.com/aeschli/typescript-vscode-sh-plugin/issues/22)

File: src/server/protocol.ts (hunk 3 above)
Comment (orta) [reply src/server/protocol.ts:873]: No, I don't think so - the data looks like:

>          * @returns a number array encoded as triples of [start, length, ClassificationType, ...].

**Comment by orta** [comment]
Just updating the baselines, I'll also submit this to the 4.2 branch 👍🏻

File: src/server/protocol.ts [hunk 4]
```diff
@@ -868,6 +866,18 @@ namespace ts.server.protocol {
...
+    export interface EncodedSemanticClassificationsResponse extends Response {
+        body?: EncodedSemanticClassificationsResponseBody
+    }
+
+    /**
+     * Implementation response message. Gives series of text spans depending on the format ar.
+     */
+    export interface EncodedSemanticClassificationsResponseBody {
+        endOfLineState: EndOfLineState;
+        spans: ClassifiedSpan[] | number[];
```
Comment (sheetalkamat) [inline src/server/protocol.ts:879]: This isnt correct.  The response currently from `getEncodedSemanticClassifications` returns type as `ts.Classifications` where spans is `number[]`. 
More over you are referencing `ClassifiedSpan` here which is only in services and not in protocol. 
This doesnt correctly reflect what we return.

Comment (sheetalkamat) [inline src/server/protocol.ts:879]: This isnt correct.  The response currently from `getEncodedSemanticClassifications` returns type as `ts.Classifications` where spans is `number[]`. 
More over you are referencing `ClassifiedSpan` here which is only in services and not in protocol. 
This doesnt correctly reflect what we return.

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

File: src/server/protocol.ts (hunk 4 above)
Comment (orta) [reply src/server/protocol.ts:879]: Dang, I thought I reverted that, that's what I get for shipping late at night, Running baselines to update now and will double check before I ask for a re-review! Thanks

Comment (orta) [reply src/server/protocol.ts]: OK, that looks good until we add some sort of wild syntax like `spans: [...[number, number, ClassificationType][] ]` heh

**Review by sheetalkamat — APPROVED** [review]

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by orta** [comment]
Cool, I'm going to replicate this for the 4.2 branch now

**Pull request merged**